runners/direct-java/src/test/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactoryTest.java

```diff
@@ -434,9 +432,13 @@ public boolean startImpl() throws IOException {
 
     @Override
     public boolean advanceImpl() throws IOException {
-      if (index == sleepIndex) {
+      // Sleep before the sleep/split index is claimed so long as it will be claimed
+      if (index + 1 == sleepIndex && sleepIndex < getCurrentSource().elems.length) {
         try {
           dynamicallySplit.await();
+          while (initialSource.equals(getCurrentSource())) {
+            // Spin until the current source is updated
```

Comment:
Note sure I've grokked everything, but I'd love an active way to receive this message. Perhaps override `splitAtFraction` and send the signal after the call to `super.splitAtFraction` completes?